File: Baekjoon/분할 정복/MinjuCho/2630.cpp
// 문제 풀이 링크: https://velog.io/@cuppizza/백준-2630-색종이-만들기-파이썬-C
#include <iostream>
#include <vector>

using namespace std;

vector<vector<int>> paper;
int counts[2] = { 0, 0 };

void div_half(int x, int y, int n)
{
	if (n == 1)
	{
		++counts[paper[y][x]];
	}
	else
	{
		int pivot = paper[y][x];
		bool isPivot = true;
		for (int i = y; i < y + n; ++i)
		{
			for (int j = x; j < x + n; ++j)
			{
				if (paper[i][j] != pivot)
				{
					isPivot = false;
					break;
				}
			}
			if (!isPivot)
			{
				break;
			}
		}
		if (isPivot)
		{
			++counts[paper[y][x]];
			return;
		}
		else
		{
			div_half(x, y, n / 2);
			div_half(x + n / 2, y, n / 2);
			div_half(x, y + n / 2, n / 2);
			div_half(x + n / 2, y + n / 2, n / 2);
		}
	}
}

int main(void)
{
	int N;
	cin >> N;
	for (int i = 0; i < N; ++i)
	{
		vector<int> input(N);
		for (int j = 0; j < N; ++j)
		{
			cin >> input[j];
		}
		paper.push_back(input);
	}
	div_half(0, 0, N);
	cout << counts[0] << endl << counts[1] << endl;
}

File: Baekjoon/분할 정복/MinjuCho/2630.py
# 문제 풀이 링크: https://velog.io/@cuppizza/백준-2630-색종이-만들기-파이썬-C
import sys
input = sys.stdin.readline
N = int(input())
paper = []
for _ in range(N):
    paper.append(list(map(int, input().split())))
counts = [0, 0]

def div_half(x, y, n):
    half = int(n/2)
    if n == 1:
        counts[paper[y][x]]+=1
        return
    else:
        pivot = paper[y][x]
        isPivot = True
        for i in range(y, y+n):
            for j in range(x, x+n):
                if paper[i][j] != pivot:
                    isPivot = False
                    break
            if isPivot == False:
                break
        if isPivot == False:
            div_half(x, y, half)
            div_half(x+half, y, half)
            div_half(x, y+half, half)
            div_half(x+half, y+half, half)
        else:
            counts[paper[y][x]]+=1
            return

div_half(0,0,N)
print(counts[0])
print(counts[1])
